window.open with noopener shouldn't create page in opener process

achristensen07 opened the pull request:
#### 9fcc0a6c7237291ae714b3a67fce464c96997062
<pre>
window.open with noopener shouldn&apos;t create page in opener process
<a href="https://bugs.webkit.org/show_bug.cgi?id=279031">https://bugs.webkit.org/show_bug.cgi?id=279031</a>
<a href="https://rdar.apple.com/135438724">rdar://135438724</a>

Reviewed by Charlie Wolfe.

We used to make a page in the opener process then navigate it and immediately discard it.
That doesn&apos;t play well with site isolation&apos;s process accounting.  JS can&apos;t observe this
because window.open returns null in such a case.

In order to implement this, I need to only call API::PageConfiguration::setRelatedPage
if there is actually an opener, and I need to remove the now duplicate checks that verify
the API client is using the provided WKWebViewConfiguration by checking the related page.
There is already another check that verifies using the opener info.

The referrer set in WebCore::createWindow needs to check WindowFeatures::wantsNoReferrer
like the one in LocalDOMWindow::createWindow.  This referrer was never used because until
now LocalDOMWindow::createWindow made a second FrameLoadRequest with the correct referrer,
but now the referrer in WebCore::createWindow needs to be correct because that is what is
sent to another process now to be the referrer of the initial request of the new page.

A few layout tests in http/tests/dom count the number of windows in the process that is
finishing the test.  Since one of the windows is now opening in another process in modern
WebKit, the count goes from 3 to 2.  WebKitLegacy test expectations are unchanged for
those three tests.  Same with a few other tests that now have an error printed in multiple
processes.

* LayoutTests/http/tests/dom/noopener-window-not-targetable-expected.txt:
* LayoutTests/http/tests/dom/noopener-window-not-targetable.html:
* LayoutTests/http/tests/dom/noopener-window-not-targetable2-expected.txt:
* LayoutTests/http/tests/dom/noopener-window-not-targetable2.html:
* LayoutTests/http/tests/dom/noreferrer-window-not-targetable-expected.txt:
* LayoutTests/http/tests/dom/noreferrer-window-not-targetable.html:
* LayoutTests/http/tests/download/sandboxed-iframe-download-not-allowed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/cross-origin-opener-policy/coop-coep-sandbox.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/cross-origin-opener-policy/coop-sandbox.https-expected.txt:
* LayoutTests/platform/mac-wk1/http/tests/dom/noopener-window-not-targetable-expected.txt: Copied from LayoutTests/http/tests/dom/noopener-window-not-targetable-expected.txt.
* LayoutTests/platform/mac-wk1/http/tests/dom/noopener-window-not-targetable2-expected.txt: Copied from LayoutTests/http/tests/dom/noopener-window-not-targetable2-expected.txt.
* LayoutTests/platform/mac-wk1/http/tests/dom/noreferrer-window-not-targetable-expected.txt: Copied from LayoutTests/http/tests/dom/noreferrer-window-not-targetable-expected.txt.
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::createWindow):
* Source/WebKit/Shared/NavigationActionData.h:
* Source/WebKit/UIProcess/API/APINavigationAction.h:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageLoadURLWithUserData):
(WKPageLoadURLRequestWithUserData):
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::createNewPage):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::finishAttachingToWebProcess):
(WebKit::WebPageProxy::loadRequest):
(WebKit::WebPageProxy::createNewPage):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::create):
(WebKit::m_textAnimationController):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
((ProcessSwap, SameSiteWindowOpenNoOpener)):
((ProcessSwap, SameSiteWindowOpenWithNameNoOpener)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, OpenWithNoopener)):

Canonical link: <a href="https://commits.webkit.org/285310@main">https://commits.webkit.org/285310@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/758fdffc37154c90b356ce57eeded1bcdcdf7c7a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72094 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51521 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24894 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76259 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23303 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74209 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59329 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23123 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56868 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15386 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75161 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46713 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62114 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37306 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43374 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19588 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21653 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65279 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19948 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77939 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16338 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19110 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65343 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_anchor_download_block_downloads.tentative.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16381 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62137 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64601 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12809 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6460 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11083 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47312 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2097 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48381 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49668 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48125 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -1934,7 +1934,7 @@ WebProcessProxy& WebPageProxy::ensureRunningProcess()
 }
 
-RefPtr<API::Navigation> WebPageProxy::loadRequest(ResourceRequest&& request, ShouldOpenExternalURLsPolicy shouldOpenExternalURLsPolicy, IsPerformingHTTPFallback isPerformingHTTPFallback, API::Object* userData)
+RefPtr<API::Navigation> WebPageProxy::loadRequest(WebCore::ResourceRequest&& request, ShouldOpenExternalURLsPolicy shouldOpenExternalURLsPolicy, IsPerformingHTTPFallback isPerformingHTTPFallback, std::unique_ptr<NavigationActionData> lastNavigationAction, API::Object* userData)
```
rvalue for lastNavigationAction?

achristensen07 committed:
window.open with noopener shouldn't create page in opener process
https://bugs.webkit.org/show_bug.cgi?id=279031
rdar://135438724

Reviewed by Charlie Wolfe.

We used to make a page in the opener process then navigate it and immediately discard it.
That doesn't play well with site isolation's process accounting.  JS can't observe this
because window.open returns null in such a case.

In order to implement this, I need to only call API::PageConfiguration::setRelatedPage
if there is actually an opener, and I need to remove the now duplicate checks that verify
the API client is using the provided WKWebViewConfiguration by checking the related page.
There is already another check that verifies using the opener info.

The referrer set in WebCore::createWindow needs to check WindowFeatures::wantsNoReferrer
like the one in LocalDOMWindow::createWindow.  This referrer was never used because until
now LocalDOMWindow::createWindow made a second FrameLoadRequest with the correct referrer,
but now the referrer in WebCore::createWindow needs to be correct because that is what is
sent to another process now to be the referrer of the initial request of the new page.

A few layout tests in http/tests/dom count the number of windows in the process that is
finishing the test.  Since one of the windows is now opening in another process in modern
WebKit, the count goes from 3 to 2.  WebKitLegacy test expectations are unchanged for
those three tests.  Same with a few other tests that now have an error printed in multiple
processes.

* LayoutTests/http/tests/dom/noopener-window-not-targetable-expected.txt:
* LayoutTests/http/tests/dom/noopener-window-not-targetable.html:
* LayoutTests/http/tests/dom/noopener-window-not-targetable2-expected.txt:
* LayoutTests/http/tests/dom/noopener-window-not-targetable2.html:
* LayoutTests/http/tests/dom/noreferrer-window-not-targetable-expected.txt:
* LayoutTests/http/tests/dom/noreferrer-window-not-targetable.html:
* LayoutTests/http/tests/download/sandboxed-iframe-download-not-allowed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/cross-origin-opener-policy/coop-coep-sandbox.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/cross-origin-opener-policy/coop-sandbox.https-expected.txt:
* LayoutTests/platform/mac-wk1/http/tests/dom/noopener-window-not-targetable-expected.txt: Copied from LayoutTests/http/tests/dom/noopener-window-not-targetable-expected.txt.
* LayoutTests/platform/mac-wk1/http/tests/dom/noopener-window-not-targetable2-expected.txt: Copied from LayoutTests/http/tests/dom/noopener-window-not-targetable2-expected.txt.
* LayoutTests/platform/mac-wk1/http/tests/dom/noreferrer-window-not-targetable-expected.txt: Copied from LayoutTests/http/tests/dom/noreferrer-window-not-targetable-expected.txt.
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::createWindow):
* Source/WebKit/Shared/NavigationActionData.h:
* Source/WebKit/UIProcess/API/APINavigationAction.h:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageLoadURLWithUserData):
(WKPageLoadURLRequestWithUserData):
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::createNewPage):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::finishAttachingToWebProcess):
(WebKit::WebPageProxy::loadRequest):
(WebKit::WebPageProxy::createNewPage):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::create):
(WebKit::m_textAnimationController):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
((ProcessSwap, SameSiteWindowOpenNoOpener)):
((ProcessSwap, SameSiteWindowOpenWithNameNoOpener)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, OpenWithNoopener)):

Canonical link: https://commits.webkit.org/285310@main
